src/main/java/io/supertokens/totp/Totp.java

```diff
@@ -0,0 +1,365 @@
+            }
+        }
+
+        return;
+
+    }
+
+    public static boolean verifyDevice(Main main, String userId, String deviceName, String code)
+            throws StorageQueryException, TotpNotEnabledException, UnknownDeviceException, InvalidTotpException,
+            LimitReachedException {
+        // Here boolean return value tells whether the device has been
+        // newly verified (true) OR it was already verified (false)
+
+        TOTPSQLStorage totpStorage = StorageLayer.getTOTPStorage(main);
+        TOTPDevice matchingDevice = null;
+
+        // Here one race condition is that the same device
+        // is to be verified twice in parallel. In that case,
+        // both the API calls will return true, but that's okay.
+
+        // Check if the user has any devices:
+        TOTPDevice[] devices = totpStorage.getDevices(userId);
+        if (devices.length == 0) {
+            throw new TotpNotEnabledException();
+        }
+
+        // Check if the requested device exists:
+        for (TOTPDevice device : devices) {
+            if (device.deviceName.equals(deviceName)) {
+                matchingDevice = device;
+                if (device.verified) {
+                    return false; // Was already verified
+                }
+                break;
+            }
+        }
+
+        // No device found:
+        if (matchingDevice == null) {
+            throw new UnknownDeviceException();
+        }
+
+        checkAndStoreCode(main, totpStorage, userId, new TOTPDevice[] { matchingDevice }, code);
+        // Will reach here only if the code is valid:
+        totpStorage.markDeviceAsVerified(userId, deviceName);
+        return true; // Newly verified
+    }
+
+    public static void verifyCode(Main main, String userId, String code, boolean allowUnverifiedDevices)
+            throws StorageQueryException, TotpNotEnabledException, InvalidTotpException,
+            LimitReachedException {
+        TOTPSQLStorage totpStorage = StorageLayer.getTOTPStorage(main);
+
+        // Check if the user has any devices:
+        TOTPDevice[] devices = totpStorage.getDevices(userId);
+        if (devices.length == 0) {
+            throw new TotpNotEnabledException();
+        }
+
+        // Filter out unverified devices:
+        if (!allowUnverifiedDevices) {
+            devices = Arrays.stream(devices).filter(device -> device.verified).toArray(TOTPDevice[]::new);
+        }
+
+        checkAndStoreCode(main, totpStorage, userId, devices, code);
+    }
+
+    /** Delete device and also delete the user if deleting the last device */
+    public static void removeDevice(Main main, String userId, String deviceName)
+            throws StorageQueryException, UnknownDeviceException, TotpNotEnabledException,
+            StorageTransactionLogicException {
+        TOTPSQLStorage storage = StorageLayer.getTOTPStorage(main);
+
+        try {
+            storage.startTransaction(con -> {
```

Comment:
for psql and mysql plugins, we need to change startTransaction to also account for retries based on delete serialisable error. No need to do anything about this comment in this core PR, we will revisit this when working on the psql and mysql plugins

Reply:
for postgresql: https://github.com/supertokens/supertokens-postgresql-plugin/blob/master/src/main/java/io/supertokens/storage/postgresql/Start.java#L220

Reply:
So I think we don't need to change anything because it returns a `concurrent update` error which is already covered.

```
ERROR:  could not serialize access due to concurrent update
CONTEXT:  SQL statement "DELETE FROM ONLY "public"."totp_used_codes" WHERE $1::pg_catalog.text OPERATOR(pg_catalog.=) "user_id"::pg_catalog.text"
```

Here's what I did (t1 = terminal1, t2 = terminal2)

```
t1: begin transaction isolation level serializable;
t2: begin transaction isolation level serializable;
t1: SELECT * FROM totp_used_codes where user_id = 'user1' FOR UPDATE;
t1: INSERT INTO totp_used_codes (user_id, code, is_valid, created_time_ms, expiry_time_ms)
VALUES ('user1', '12345x', true, 123456781, 123456789);
t2: DELETE FROM totp_users where user_id = 'user1';

Now t2 kept waiting, until I did:
t1: COMMIT;

Now t2 throws the error that I mentioned above
```

Reply:
```
t1: SET TRANSACTION ISOLATION LEVEL SERIALIZABLE;
t2: SET TRANSACTION ISOLATION LEVEL SERIALIZABLE;

t1: START TRANSACTION;
t2: START TRANSACTION;
t2: DELETE FROM totp_users where user_id = 'user1';
t1: SELECT * FROM totp_used_codes where user_id = 'user1' FOR UPDATE;

t1: INSERT INTO totp_used_codes (user_id, code, is_valid, created_time_ms, expiry_time_ms)
    -> VALUES ('user1', '12345x', true, 123456781, 123456789);
  
t1 Gives error:
ERROR 1213 (40001): Deadlock found when trying to get lock; try restarting transaction
```

Reply:
This is not the right thing to run. We want to see the error failed post commit and it had the message that unable to seriable due to delete or something like that

Reply:
We have fixed and updated it.